src/ppx-metaquot/fun_ty.mli

```diff
@@ -0,0 +1,143 @@
+(* This file is part of Learn-OCaml.
+ *
+ * Copyright (C) 2019 OCaml Software Foundation.
+ * Copyright (C) 2016-2018 OCamlPro.
+ *
+ * Learn-OCaml is distributed under the terms of the MIT license. See the
+ * included LICENSE file for details. *)
+
+(** [Fun_ty] is used by [Test_lib] to implement n-ary graders.
+
+    This module provides two GADTs: [args], representing arguments of
+    n-ary functions as nonempty heterogeneous lists, and [fun_ty],
+    representing function types in a more explicit way than ['a Ty.ty].
+
+    This module also serves as a runtime library for the customized
+    [ppx_metaquot], so that one may write:
+
+    [let p = [%funty: int -> bool]], and get:
+
+    [val p : ((int -> bool) Ty.ty, int -> unit, bool) fun_ty = <abstr>] *)
+
+(** GADT for arguments of n-ary functions, implemented as nonempty
+    heterogeneous lists.
+
+    Usage: [arg 3 @@ arg "word" @@ last false]
+
+    Alternatively: [3 @: "word" @:!! false] *)
+type ('arrow, 'uarrow, 'ret) args
+
+(** [last e], or equivalently [!! e], builds a one-element argument list *)
+val last :
+  'a ->
+  ('a -> 'ret, 'a -> unit, 'ret) args
+
+(** [arg a l], or equivalently [a @: l], adds [a] in front of the
+   argument list [l] *)
+val arg :
+  'a ->
+  ('ar -> 'row, 'ar -> 'urow, 'ret) args ->
+  ('a -> 'ar -> 'row, 'a -> 'ar -> 'urow, 'ret) args
+
+(** Helper notation for [last] *)
+val (!!) :
+  'a ->
+  ('a -> 'ret, 'a -> unit, 'ret) args
+
+(** Helper notation for [arg] *)
+val (@:) :
+  'a ->
+  ('ar -> 'row, 'ar -> 'urow, 'ret) args ->
+  ('a -> 'ar -> 'row, 'a -> 'ar -> 'urow, 'ret) args
+
+(** [a @:!! l] is another notation for [a @: !! l] (with a space) *)
+val (@:!!) :
+  'a -> 'b ->
+  ('a -> 'b -> 'ret, 'a -> 'b -> unit, 'ret) args
+
+(** [apply f l] applies a n-ary function [f] to the arguments from [l] *)
+val apply :
+  ('ar -> 'row) -> ('ar -> 'row, 'ar -> 'urow, 'ret) args ->
+  'ret
+
+(** GADT for function types.
+
+    Given an arrow type ['a -> 'row], the following construct provides
+    a more precise representation of this function type than
+    [[%ty: 'a -> 'row] : ('a -> 'row) Ty.ty]:
+
+    [[%funty: 'a -> 'row] : (('a -> 'row) Ty.ty, 'a -> 'urow, 'ret) fun_ty]
+
+    In particular, the codomain type ['ret] is made explicit, so that
+    if ['row = 'b -> 'c], we get ['urow = 'b -> unit] and ['ret = 'c].
+
+    Usage: [arg_ty [%ty: int] @@ arg_ty [%ty: string] @@
+            last_ty [%ty: bool] [%ty: unit]]
```

Comment:
@yurug OK, but even if the current naming is not the best possible I guess that we should maybe keep it as is for backward compatibility, because this material was already exported by the interface `Test_lib` since the first public commit of learn-ocaml [`23aa830`](https://github.com/ocaml-sf/learn-ocaml/blob/23aa830d4b00bf4b9226aba1159564585d957de6/src/grader/test_lib.mli#L256-L266), and these two functions `arg_ty` and `last_ty` are the most important functions related to n-ary function types (they are wrappers for the constructors [`Last_ty` and `Arg_ty`](https://github.com/ocaml-sf/learn-ocaml/blob/23aa830d4b00bf4b9226aba1159564585d957de6/src/grader/test_lib.ml#L1025-L1026)).

So I propose to keep this naming as is… Do you agree?